ddpui/core/ai/tools/catalog.py

```
@@ -0,0 +1,64 @@
+def check_schema(ctx: RunContext, schema: str) -> None:
+    if schema not in ctx.allowed_schemas:
+        raise ToolInputError(
+            f"Schema '{schema}' is not available. Available schemas: {sorted(ctx.allowed_schemas)}"
+        )
+
+
+def list_table_names(ctx: RunContext, schema: str) -> dict[str, int | None]:
+    """{table_name: approx_rows} for a validated schema, via dialect catalog SQL."""
+    check_schema(ctx, schema)
+    if ctx.dialect == "bigquery":
+        sql = f"SELECT table_id AS table_name, row_count AS approx_rows FROM `{schema}.__TABLES__`"
```

Ruff (0.15.21)

[error] Possible SQL injection vector through string-based query construction

(S608)